### PR TITLE
chore(atomic): Make Postgres transactions atomic

### PR DESCRIPTION
**Pull request by thoven87**
Previously, jobs would get queued. They would never get processed until the server is restarted.
This is PR fixes this behavior with transactions and Postgres `FOR UPDATE SKIP LOCKED` [Doc](https://www.postgresql.org/docs/current/sql-select.html)

This PR was tested on Postgres 16 with swift 6.

### CODE REVIEW

File: Sources/JobsPostgres/PostgresJobsQueue.swift [hunk 1]
```diff
@@ -302,6 +310,18 @@ extension PostgresJobQueue {
...
+
+    private func startTransaction(connection: PostgresConnection) async throws {
+        try await connection.query("BEGIN;", logger: self.logger)
+    }
+
+    private func commitTransaction(connection: PostgresConnection) async throws {
+        try await connection.query("COMMIT;", logger: self.logger)
+    }
+
+    private func rollbackTransaction(connection: PostgresConnection) async throws {
```
Comment (adam-fowler) [inline Sources/JobsPostgres/PostgresJobsQueue.swift:322]: Could we not combine these three functions into one function `PostgresClient.withTransaction(logger:)` eg
```swift
extension PostgresClient {
    func withTransaction<Value>(logger: Logger, _ process: (PostgresConnection) async throws -> Value) async throws -> Value {
        try await withConnection { connection in
            do {
                try await connection.query("BEGIN;", logger: logger)
                let value = try await process(connection)
                try await connection.query("COMMIT;", logger: logger)
                return value
            } catch {
                try await connection.query("ROLLBACK;", logger: logger)
                throw error
            }
        }
    }
}
```

File: Sources/JobsPostgres/Migrations/CreateJobQueue.swift [hunk 2]
```diff
@@ -22,15 +22,15 @@ struct CreateJobQueue: PostgresMigration {
             """
             CREATE TABLE IF NOT EXISTS _hb_pg_job_queue (
                 job_id uuid PRIMARY KEY,
-                createdAt timestamp with time zone
+                created_at timestamp with time zone
```
Comment (adam-fowler) [inline Sources/JobsPostgres/Migrations/CreateJobQueue.swift:25]: Can we remove the variable renaming. If we are going to rename the variables we should do it in another PR

File: Sources/JobsPostgres/Migrations/CreateJobs.swift [hunk 3]
```diff
@@ -24,11 +24,18 @@ struct CreateJobs: PostgresMigration {
                 id uuid PRIMARY KEY,
                 job bytea,
                 status smallint,
-                lastModified TIMESTAMPTZ DEFAULT CURRENT_TIMESTAMP
+                last_modified TIMESTAMPTZ DEFAULT NOW()
```
Comment (adam-fowler) [inline Sources/JobsPostgres/Migrations/CreateJobs.swift:27]: Can we remove the variable renaming. If we are going to rename the variables we should do it in another PR

File: Sources/JobsPostgres/Migrations/CreateJobs.swift [hunk 4]
```diff
@@ -24,11 +24,18 @@ struct CreateJobs: PostgresMigration {
...
             )     
             """,
             logger: logger
         )
+        try await connection.query(
```
Comment (adam-fowler) [inline Sources/JobsPostgres/Migrations/CreateJobs.swift:32]: Is this to improve the performance of the queries called from `onInit`?

**Review by adam-fowler — CHANGES_REQUESTED** [review]

File: Sources/JobsPostgres/Migrations/CreateJobs.swift (hunk 4 above)
Comment (thoven87) [reply Sources/JobsPostgres/Migrations/CreateJobs.swift:32]: Yes, it is.

File: Sources/JobsPostgres/PostgresJobsQueue.swift (hunk 1 above)
Comment (thoven87) [reply Sources/JobsPostgres/PostgresJobsQueue.swift:322]: That's brilliant! I will update the PR.

**Comment by thoven87** [comment]
Is looks like running `./scripts/validate.sh` changes a lot of the code format and move internal init to non internal. I suspect, the build will fail once enabled. @adam-fowler

File: Sources/JobsPostgres/PostgresJobsQueue.swift (hunk 1 above)
Comment (thoven87) [reply Sources/JobsPostgres/PostgresJobsQueue.swift]: I think this util function should extracted and move to PostgresNIO.

File: Sources/JobsPostgres/PostgresJobsQueue.swift [hunk 5]
```diff
@@ -160,9 +160,10 @@ public final class PostgresJobQueue: JobQueueDriver {
 
     func popFirst() async throws -> QueuedJob<JobID>? {
         do {
-            return try await self.client.withConnection { connection -> QueuedJob? in
+            return try await self.client.withTransaction(logger: self.logger) { connection -> QueuedJob? in
```
Comment (adam-fowler) [inline Sources/JobsPostgres/PostgresJobsQueue.swift:163]: There's one issue with this. At the bottom of this `withTransaction` block if the job fails to decode, it throws a `JobQueueError.decodeJobFailed` error which will then cancel the whole of the transaction. Thus the job is never popped off the queue and flagged as failed and the next time it comes to pop a job off the queue it'll do exactly the same thing with the same job because it is still at the top of the queue.

We could return a `Result<QueuedJob<JobID>, Error>` from the `withTransaction` block. And if it is an error throw the error once outside of the `withTransaction` block

**Review by adam-fowler — CHANGES_REQUESTED** [review]
Looks good outside of one issue

File: Sources/JobsPostgres/PostgresJobsQueue.swift (hunk 5 above)
Comment (thoven87) [reply Sources/JobsPostgres/PostgresJobsQueue.swift:163]: Understood! Eventually, I wanted to move the table to [this schema](https://github.com/hummingbird-project/hummingbird-postgres/issues/19) which would afford us to drop jobs that has failed for N number of times. I will update the PR with the above suggestion.

File: Sources/JobsPostgres/PostgresJobsQueue.swift [hunk 6]
```diff
@@ -158,11 +158,12 @@ public final class PostgresJobQueue: JobQueueDriver {
     /// shutdown queue once all active jobs have been processed
     public func shutdownGracefully() async {}
 
-    func popFirst() async throws -> QueuedJob<JobID>? {
+    func popFirst() async throws -> Result<QueuedJob<JobID>?, Error> {
```
Comment (adam-fowler) [inline Sources/JobsPostgres/PostgresJobsQueue.swift:161]: It seems a little odd to have a throwing function return a `Result` ie it has two ways to report errors. I would still have it return the QueuedJob<JobID>, but at the bottom of the function return `try result.get()`

```swift
let result = try await self.client.withTransaction {
    ...
}
return try result.get()
```

File: Sources/HummingbirdPostgres/Migration.swift [hunk 7]
```diff
@@ -18,7 +18,7 @@ import PostgresNIO
 /// Protocol for a database migration
 ///
 /// Requires two functions one to apply the database migration and one to revert it.
-public protocol PostgresMigration {
+public protocol PostgresMigration: Sendable {
```
Comment (adam-fowler) [inline Sources/HummingbirdPostgres/Migration.swift:21]: Is there a reason this has Sendable conformance?

Comment (thoven87) [reply Sources/HummingbirdPostgres/Migration.swift:21]: The validate script added the Sendable conformance.

Comment (thoven87) [reply Sources/HummingbirdPostgres/Migration.swift:21]: > Is there a reason this has Sendable conformance?

Same as the internal init removal.

Comment (adam-fowler) [reply Sources/HummingbirdPostgres/Migration.swift:21]: oh that's odd. Is that SwiftFormat trying to be overly clever? What version of SwiftFormat do you have?

Comment (thoven87) [reply Sources/HummingbirdPostgres/Migration.swift:21]: > oh that's odd. Is that SwiftFormat trying to be overly clever? What version of SwiftFormat do you have?

```sh swiftformat --version
0.54.3
```
It looks like version 0.54.3

**Review by adam-fowler — APPROVED** [review]
I removed the Sendable on the PostgresMigration protocol
So everything is good to go in

**Comment by thoven87** [comment]
> I removed the Sendable on the PostgresMigration protocol So everything is good to go in

Thanks @adam-fowler Does the validate.sh file need an update? 

I don't have merge privileges, please merge at your own convenience.

Thanks again for the help. I look forward to contributing more to the project.
